#### Chi Square Computation with observed and expected frequencies given

In [1]:
observed = [38,38,20,4]
expected = [44,41,10,5]
significance = 0.05

def chi_square_observed_expected(observed_frequencies,expected_frequencies,alpha):
    from scipy.stats import chi2
    
    observed_chi_square = round(sum([round((observed_frequencies[i]-expected_frequencies[i])**2/expected_frequencies[i],3) for i in range(len(expected_frequencies))]),3)
    print(f'observed_chi_square = {observed_chi_square}')

    df = len(observed_frequencies)-1
    print(f'df = {df}')

    chi_crit = round(chi2.ppf(1-alpha,df),3)
    print(f'chi_crit = {chi_crit}')

    if observed_chi_square > chi_crit:
        print(f'observed_chi_square = {observed_chi_square} is GREATER than the critical value = {chi_crit}. We will REJECT the Null Hypothesis')
    else:
        print(f'observed_chi_square = {observed_chi_square} is LESS than the critical value = {chi_crit}. We will RETAIN the Null Hypothesis')
chi_square_observed_expected(observed,expected,significance)

observed_chi_square = 11.238
df = 3
chi_crit = 7.815
observed_chi_square = 11.238 is GREATER than the critical value = 7.815. We will REJECT the Null Hypothesis


#### Chi Square Computation

In [2]:
def chi_square_computation(list,alpha):
    from scipy.stats import chi2
    new_list = [list[j][i] for i in range(len(list[0])) for j in range(len(list))]
    print(f'new_list = {new_list}')

    row_total = [sum(new_list[i:i+len(list)]) for i in range(0,len(new_list),len(list))]
    print(f'row_total = {row_total}')

    observed_frequencies = []
    [observed_frequencies.extend(item) for item in list]
    
    column_total = [sum(item) for item in list]
    print(f'column_total = {column_total}')

    grand_total = sum(row_total)
    print(f'grand_total = {grand_total}')

    print(f'{len(column_total)}x{len(row_total)} dimension of table')
    df = (len(column_total)-1)*(len(row_total)-1)
    print(f'df = ({len(column_total)}-1) x ({len(row_total)}-1)')
    print(f'df = {df}')

    chi_crit = round(chi2.ppf(1-alpha,df),3)
    print(f'chi_crit = {chi_crit}')
    
    expected_frequencies = [round((row_total[j]*column_total[i])/grand_total,3) for i in range(len(column_total)) for j in range(len(row_total))]
    print(f'observed_frequencies = {observed_frequencies}')
    print(f'expected_frequencies = {expected_frequencies}')

    observed_chi_square = round(sum([round((observed_frequencies[i]-expected_frequencies[i])**2/expected_frequencies[i],3) for i in range(len(expected_frequencies))]),3)
    print(f'observed_chi_square = {observed_chi_square}')

    if observed_chi_square > chi_crit:
        print(f'At {alpha} significance level, the observed value of Chi Square = {observed_chi_square} is GREATER than the critical value = {chi_crit}. We will REJECT the Null Hypothesis')
    else:
        print(f'At {alpha} significance level, the observed value of Chi Square = {observed_chi_square} is LESS than the critical value = {chi_crit}. We will RETAIN the Null Hypothesis')

chi_square_computation([[375,379,46],[367,336,37],[357,362,41]],0.05)
# chi_square_computation([[375,379,46],[367,336,37]],0.05)
# chi_square_computation([[375,379,46],[367,336,37],[357,362,41],[1,2,3],[4,5,6]],0.05)

new_list = [375, 367, 357, 379, 336, 362, 46, 37, 41]
row_total = [1099, 1077, 124]
column_total = [800, 740, 760]
grand_total = 2300
3x3 dimension of table
df = (3-1) x (3-1)
df = 4
chi_crit = 9.488
observed_frequencies = [375, 379, 46, 367, 336, 37, 357, 362, 41]
expected_frequencies = [382.261, 374.609, 43.13, 353.591, 346.513, 39.896, 363.148, 355.878, 40.974]
observed_chi_square = 1.627
At 0.05 significance level, the observed value of Chi Square = 1.627 is LESS than the critical value = 9.488. We will RETAIN the Null Hypothesis


In [3]:
obs = [25,10,5,25,10,15]
exp = [round((1/6)*num,3) for num in obs]
print(exp)
sig = 0.05
chi_square_observed_expected(obs,exp,sig)

[4.167, 1.667, 0.833, 4.167, 1.667, 2.5]
observed_chi_square = 374.965
df = 5
chi_crit = 11.07
observed_chi_square = 374.965 is GREATER than the critical value = 11.07. We will REJECT the Null Hypothesis


In [4]:
# From Witte, p.372, Table 19.3
data = [[39,21],[30,40],[51,19]]
sig = 0.05
chi_square_computation(data,sig)

new_list = [39, 30, 51, 21, 40, 19]
row_total = [120, 80]
column_total = [60, 70, 70]
grand_total = 200
3x2 dimension of table
df = (3-1) x (2-1)
df = 2
chi_crit = 5.991
observed_frequencies = [39, 21, 30, 40, 51, 19]
expected_frequencies = [36.0, 24.0, 42.0, 28.0, 42.0, 28.0]
observed_chi_square = 14.019
At 0.05 significance level, the observed value of Chi Square = 14.019 is GREATER than the critical value = 5.991. We will REJECT the Null Hypothesis


### From Witte p.376
Progress Check *19.4 An investigator suspects that there might be a relationship, possibly based on genetic factors, between hair color and susceptibility to poison oak. Three hundred volunteer subjects are exposed to a small amount of poison oak and then classified according to their susceptibility (rash or no rash) and their hair color (red, blond, brown, or black), yielding the following frequencies:

HAIR COLOR AND SUSCEPTIBILITY TO POISON OAK HAIR COLOR
\begin{array}{ccc}
\text{SUSCEPTIBILITY} & \text{RED} & \text{BLOND} & \text{BROWN} & \text{BLACK} & \text{TOTAL}\\
\hline
Rash & 10 & 30 & 60 & 80 & 180 \\
No~rash & 20 & 30 & 30 & 40 & 120 \\
TOTAL & 30 & 60 & 90 & 120 & 300\\
\end{array}

(a) Test the null hypothesis at the .01 level of significance.

(b) Specify the approximate p-value for this test result.

In [6]:
data1 = [[10,20],[30,60],[60,30],[80,40]]
sig1 = 0.01
chi_square_computation(data1,sig1)

new_list = [10, 30, 60, 80, 20, 60, 30, 40]
row_total = [180, 150]
column_total = [30, 90, 90, 120]
grand_total = 330
4x2 dimension of table
df = (4-1) x (2-1)
df = 3
chi_crit = 11.345
observed_frequencies = [10, 20, 30, 60, 60, 30, 80, 40]
expected_frequencies = [16.364, 13.636, 49.091, 40.909, 49.091, 40.909, 65.455, 54.545]
observed_chi_square = 34.222
At 0.01 significance level, the observed value of Chi Square = 34.222 is GREATER than the critical value = 11.345. We will REJECT the Null Hypothesis


#### PROPORTION OF EXPLAINED VARIANCE (TWO-VARIABLE $\chi^2$)

$\phi_c^2=\dfrac{\chi^2}{n(k-1)}$

$k - number~of~groups$

In [8]:
phi_squared_sub_c = 34.222/300*(2-1)
phi_squared_sub_c

0.11407333333333333